src/idlcxx/src/streamers.c

```diff
@@ -106,15 +114,172 @@ process_instance(
   if (idl_is_constr_type(type_spec) &&
       !idl_is_keyless(type_spec, pstate->flags & IDL_FLAG_KEYLIST))
     fmt = "  key_%2$s(str, %1$s);\n";
-  if (putf(&streams->key_write, fmt, name, "write")
-   || putf(&streams->key_read, fmt, name, "read")
-   || putf(&streams->key_move, fmt, name, "move")
-   || putf(&streams->key_max, fmt, name, "max"))
+
+  if (putf(&streams->key_write, fmt, accessor, "write", maximum)
+    || putf(&streams->key_read, fmt, accessor, "read", maximum)
+    || putf(&streams->key_move, fmt, accessor, "move", maximum)
+    || putf(&streams->key_max, fmt, accessor, "max", maximum))
     return IDL_RETCODE_NO_MEMORY;
 
   return IDL_RETCODE_OK;
 }
 
+static idl_retcode_t
+unroll_sequence(const idl_pstate_t* pstate,
+  struct streams* streams,
+  const idl_sequence_t* seq,
+  size_t depth,
+  const char* accessor,
+  bool is_key)
+{
+  uint32_t maximum = seq->maximum;
+
+  const char* wfmt = maximum ?
+    "  uint32_t se_%1$u = std::min<uint32_t>(uint32_t(%2$s.size()), %3$u);\n"\
+    "  %4$s(str, se_%1$u);\n"\
+    "  for (uint32_t i_%1$u = 0; i_%1$u < se_%1$u; i_%1$u++) {\n" :
+    "  uint32_t se_%1$u = uint32_t(%2$s.size());\n"\
+    "  %4$s(str, se_%1$u);\n"\
+    "  for (uint32_t i_%1$u = 0; i_%1$u < se_%1$u; i_%1$u++) {\n";
+  const char* rfmt = "  uint32_t se_%1$u;\n"\
+    "  read(str, se_%1$u);\n"\
+    "  %2$s.resize(se_%1$u);\n"\
+    "  for (uint32_t i_%1$u = 0; i_%1$u < se_%1$u; i_%1$u++) {\n";
+  const char* mfmt = "  uint32_t se_%1$u = uint32_t(%2$u);\n"\
+    "  max(str, se_%1$u);\n"\
+    "  for (uint32_t i_%1$u = 0; i_%1$u < se_%1$u; i_%1$u++) {\n";
+
+  if (putf(&streams->read, rfmt, depth, accessor)
+    || putf(&streams->write, wfmt, depth, accessor, maximum, "write")
+    || putf(&streams->move, wfmt, depth, accessor, maximum, "move")
+    || putf(&streams->max, mfmt, depth, maximum))
+    return IDL_RETCODE_NO_MEMORY;
+
+  if (is_key)
+  {
+    if (putf(&streams->key_read, rfmt, depth, accessor)
+      || putf(&streams->key_write, wfmt, depth, accessor, maximum, "write")
+      || putf(&streams->key_move, wfmt, depth, accessor, maximum, "move")
+      || putf(&streams->key_max, mfmt, depth, maximum))
+    return IDL_RETCODE_NO_MEMORY;
+  }
+
+  char* new_accessor = NULL;
+  if (idl_asprintf(&new_accessor, "%s[i_%u]", accessor, depth) < 0)
+    return IDL_RETCODE_NO_MEMORY;
```

Comment:
Better use `IDL_PRINTA` to allow for early out.

Reply:
will change to from idl_asprintf to IDL_PRINTA wherever possible